### PR TITLE
Relay partially failed htlcs when closing

### PR DESCRIPTION
**Pull request by t-bast**
If a channel closes when we've received an `UpdateFailHtlc`, signed it but not yet received our peer's revocation, we need to fail the htlc upstream.

That specific scenario was not correctly handled, resulting in upstream htlcs that were not failed which would force our upstream peer to close the channel.

In that scenario the HTLC is in the `remoteCommit` but not in the `nextRemoteCommit`.

### CODE REVIEW

**Comment by pm47** [comment]
> If a channel closes when we've received an UpdateFailHtlc, signed it but not yet received our peer's revocation, we need to fail the htlc upstream.

I don't think we can. What if the peer publishes its previous commitment (that isn't revoked) and uses the preimage?

**Comment by t-bast** [comment]
> What if the peer publishes its previous commitment 

He can't, we waited for our local commit to be confirmed ;)

**Comment by pm47** [comment]
>> What if the peer publishes its previous commitment

>He can't, we waited for our local commit to be confirmed ;)

Well then yes. Alright, you got me I didn't go past the PR description

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 1]
```diff
@@ -954,10 +954,10 @@ object Helpers {
       val remoteCommit = d.commitments.remoteCommit
       val nextRemoteCommit_opt = d.commitments.remoteNextCommitInfo.left.toOption.map(_.nextRemoteCommit)
       if (localCommit.publishableTxs.commitTx.tx.txid == tx.txid) {
-        // our commit got confirmed, so any htlc that we signed but they didn't sign will never reach the chain
-        val mostRecentRemoteCommit = nextRemoteCommit_opt.getOrElse(remoteCommit)
+        // our commit got confirmed, so any htlc that is in their commitment but not ours will never reach the chain
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:957]:
```suggestion
        // our commit got confirmed, so any htlc that is in their commitment but not in ours will never reach the chain
```
maitre_capello.jpg

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 2]
```diff
@@ -954,10 +954,10 @@ object Helpers {
...
+        val htlcsInRemoteCommit = remoteCommit.spec.htlcs ++ nextRemoteCommit_opt.map(_.spec.htlcs).getOrElse(Set.empty)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:958]: 😎 

```suggestion
        val htlcsInRemoteCommit = remoteCommit.spec.htlcs ++ nextRemoteCommit_opt.flatMap(_.spec.htlcs)
```

**Review by pm47 — COMMENTED** [review]
When I see how complex that corner case is and how simple and elegant the fix _and_ the tests are, I think we can be proud of that codebase!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 2 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: Done in https://github.com/ACINQ/eclair/pull/1706/commits/e2119a400387f34134257c5975f884a23c0c3411

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: Done in https://github.com/ACINQ/eclair/pull/1706/commits/e2119a400387f34134257c5975f884a23c0c3411

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 2 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: Actually it doesn't build...

**Review by pm47 — APPROVED** [review]
